File: engine/src/persistence/run_repository.rs
use crate::model::{Run, RunId, RunStatus, ProjectId, ServiceError};

pub trait RunRepository {
    async fn update_status(&self, run: &Run) -> Result<(), ServiceError>;
    async fn claim_next_pending(&self) -> Result<Option<Run>, ServiceError>;
}

pub struct SqlxRunRepository{
    pool: sqlx::PgPool,
}

impl SqlxRunRepository {
    pub fn new(pool: sqlx::PgPool) -> Self {
        Self { pool }
    }
}

impl RunRepository for SqlxRunRepository {
    async fn update_status(&self, run: &Run) -> Result<(), ServiceError> {
        sqlx::query("UPDATE analysis_run SET status=$1, started_at=$2, finished_at=$3, error_message=$4 WHERE id = $5")
            .bind(run.status.to_string())
            .bind(run.started_at.as_ref())
            .bind(run.finished_at.as_ref())
            .bind(run.error_message.as_ref())
            .bind(run.id.to_string())
            .execute(&self.pool)
            .await
            .map_err(|e| ServiceError::DatabaseRequestFailed(e.to_string()))?;
        Ok(())
    }

    async fn claim_next_pending(&self) -> Result<Option<Run>, ServiceError> {
        let row = sqlx::query(
            "UPDATE analysis_run
             SET status = 'processing', started_at = NOW()
             WHERE id = (
                 SELECT id FROM analysis_run
                 WHERE status = 'pending'
                 ORDER BY created_at
                 LIMIT 1
                 FOR UPDATE SKIP LOCKED
             )
             RETURNING *"
        )
            .fetch_optional(&self.pool)
            .await
            .map_err(|e| ServiceError::DatabaseRequestFailed(e.to_string()))?;

        match row {
            None => Ok(None),
            Some(row) => {
                use sqlx::Row;
                let status: String = row.get("status");
                let run = Run {
                    id: RunId::from_raw(row.get("id")),
                    project_id: ProjectId::from_raw(row.get("project_id")),
                    branch: row.get("branch"),
                    commit: row.get("commit"),
                    status: status.parse::<RunStatus>()
                        .map_err(|e| ServiceError::DatabaseRequestFailed(e.to_string()))?,
                    error_message: row.get("error_message"),
                    started_at: row.get("started_at"),
                    finished_at: row.get("finished_at"),
                };
                Ok(Some(run))
            }
        }
    }
}

File: api/src/api/repositories/run_repository.py
from ..core import get_db
from asyncpg import Connection
from fastapi import Depends
from ..models import Run, RunId


class RunRepository:

    def __init__(self, db: Connection):
        self.db = db

    async def save(self, run: Run):
        query = """
        INSERT INTO analysis_run (id, project_id, branch, commit, status, error_message)
        VALUES ($1, $2, $3, $4, $5, $6)
        ON CONFLICT (id) DO NOTHING
        """
        await self.db.execute(
            query,
            run.id,
            run.project_id,
            run.branch,
            run.commit,
            run.status.value,
            run.error_message,
        )
    
    async def get_run(self, run_id: RunId) -> Run | None:
        query = "SELECT project_id, branch, commit, status, error_message, started_at, finished_at FROM analysis_run WHERE id = $1"
        row = await self.db.fetchrow(query, run_id)
        if row is None:
            return None
        return Run.from_db_row(run_id, row)
    
    async def get_runs_by_project_id(self, project_id: str) -> list[Run]:
        query = "SELECT id, project_id, branch, commit, status, error_message, started_at, finished_at FROM analysis_run WHERE project_id = $1"
        rows = await self.db.fetch(query, project_id)
        return [Run.from_db_row(row["id"], row) for row in rows]
        

def get_run_repository(db: Connection = Depends(get_db))-> RunRepository:
    return RunRepository(db)